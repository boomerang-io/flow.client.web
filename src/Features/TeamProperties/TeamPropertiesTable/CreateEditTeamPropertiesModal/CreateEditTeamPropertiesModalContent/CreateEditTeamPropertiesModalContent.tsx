import React from "react";
import { useMutation, useQueryClient } from "react-query";
import { Formik } from "formik";
import * as Yup from "yup";
import {
  Button,
  InlineNotification,
  ModalBody,
  ModalFooter,
  TextInput,
  Toggle,
} from "@boomerang-io/carbon-addons-boomerang-react";
import { ModalFlowForm, notify, ToastNotification, Loading } from "@boomerang-io/carbon-addons-boomerang-react";
import { serviceUrl, resolver } from "Config/servicesConfig";
import { InputType, PROPERTY_KEY_REGEX, PASSWORD_CONSTANT } from "Constants";
import { PatchProperty, Property, FlowTeam } from "Types";
import { updatedDiff } from "deep-object-diff";

type Props = {
  closeModal: () => void;
  isEdit: boolean;
  property: Property;
  propertyKeys: string[];
  team: FlowTeam;
  cancelRequestRef: any;
};

function CreateEditTeamPropertiesModalContent({
  closeModal,
  isEdit,
  property,
  propertyKeys,
  team,
  cancelRequestRef,
}: Props) {
  const initialState = {
    label: property?.label ?? "",
    description: property?.description ?? "",
    key: property?.key ?? "",
    value: property?.value ?? "",
    secured: property?.type === InputType.Password ?? false,
  };
  const teamPropertiesUrl = serviceUrl.getTeamProperties({ id: team.id });
  const queryClient = useQueryClient();

  /** Add Team Property */
  const { mutateAsync: addTeamPropertyMutation, isLoading: addIsLoading, error: addError } = useMutation(
    (args: { id: string; body: Property }) => {
      const { promise, cancel } = resolver.postTeamPropertyRequest(args);
      cancelRequestRef.current = cancel;
      return promise;
    },
    {
      onSuccess: () => queryClient.invalidateQueries(teamPropertiesUrl),
    }
  );

  /** Update Team Property */
  const { mutateAsync: updateTeamPropertyMutation, isLoading: updateIsLoading, error: updateError } = useMutation(
    (args: { teamId: string; configurationId: string; body: PatchProperty }) => {
      const { promise, cancel } = resolver.patchTeamPropertyRequest(args);
      cancelRequestRef.current = cancel;
      return promise;
    },
    {
      onSuccess: () => queryClient.invalidateQueries(teamPropertiesUrl),
    }
  );

  const loading = addIsLoading || updateIsLoading;

  const handleSubmit = async (values: any) => {
    const type = values.secured ? InputType.Password : InputType.Text;
    const newTeamProperty = isEdit ? { ...values, type, id: property.id } : { ...values, type };
    delete newTeamProperty.secured;

    if (isEdit) {
      const updatedFields = updatedDiff(initialState, newTeamProperty);
      try {
        const response = await updateTeamPropertyMutation({
          teamId: team.id,
          configurationId: newTeamProperty.id,
          body: updatedFields,
        });
        notify(
          <ToastNotification
            kind="success"
            title={"Parameter Updated"}
            subtitle={`Request to update ${response.data.label} succeeded`}
            data-testid="create-update-team-prop-notification"
          />
        );
        closeModal();
      } catch (err) {}
    } else {
      try {
        const response = await addTeamPropertyMutation({ id: team.id, body: newTeamProperty });
        notify(
          <ToastNotification
            kind="success"
            title={"Parameter Created"}
            subtitle={`Request to create ${response.data.label} succeeded`}
            data-testid="create-update-team-prop-notification"
          />
        );
        closeModal();
      } catch (err) {
        //no-op
      }
    }
  };

  // Check if key contains alpahanumeric, underscore, dash, and period chars
  const validateKey = (key: any) => {
    return PROPERTY_KEY_REGEX.test(key);
  };

  return (
    <Formik
      initialValues={initialState}
      onSubmit={handleSubmit}
      validateOnMount
      validationSchema={Yup.object().shape({
        label: Yup.string().required("Enter a label"),
        key: Yup.string()
          .required("Enter a key")
          .max(128, "Key must not be greater than 128 characters")
          .notOneOf(propertyKeys || [], "Enter a unique key value for this parameter")
          .test(
            "is-valid-key",
            "Only alphanumeric, hyphen and underscore characters allowed. Must begin with a letter or underscore",
            validateKey
          ),
        value: Yup.string().when("secured", {
          is: (value: boolean) => value && isEdit,
          then: Yup.string(),
          otherwise: Yup.string().required("Enter a value"),
        }),
        description: Yup.string(),
        secured: Yup.boolean(),
      })}
    >
      {(props) => {
        const { values, touched, errors, isValid, handleChange, handleBlur, handleSubmit } = props;

        return (
          <ModalFlowForm onSubmit={handleSubmit}>
            <ModalBody aria-label="inputs">
              {loading && <Loading />}
              <TextInput
                id="key"
                labelText="Key"
                placeholder="Key"
                name="key"
                value={values.key}
                onBlur={handleBlur}
                onChange={handleChange}
                invalid={errors.key && touched.key}
                invalidText={errors.key}
              />
              <TextInput
                id="label"
                labelText="Label"
                placeholder="Label"
                name="label"
                value={values.label}
                onBlur={handleBlur}
                onChange={handleChange}
                invalid={errors.label && touched.label}
                invalidText={errors.label}
              />
              <TextInput
                id="description"
                labelText="Description (optional)"
                placeholder="Description"
                name="description"
                value={values.description}
                onBlur={handleBlur}
                onChange={handleChange}
              />
              <TextInput
                id="value"
                labelText="Value"
                placeholder={isEdit && values.secured ? PASSWORD_CONSTANT : "Value"}
                name="value"
                value={values.value}
                onBlur={handleBlur}
                onChange={handleChange}
                invalid={errors.value && touched.value}
                invalidText={errors.value}
                type={values.secured ? "password" : "text"}
                helperText={
                  isEdit && values.secured
                    ? "Secure values are stored in our database. You can provide a new value to override"
                    : null
                }
              />
              <Toggle
                id="secured-team-parameters-toggle"
                data-testid="secured-team-parameters-toggle"
                disabled={isEdit}
                labelText="Secured"
                name="secured"
                onChange={handleChange}
                orientation="vertical"
                toggled={values.secured}
                helperText="Once a parameter is created, secured state will not be able to be updated"
              />
              {addError && (
                <InlineNotification
                  lowContrast
                  kind="error"
                  subtitle={"Request to create parameter failed"}
                  title={"Something's Wrong"}
                  data-testid="create-update-team-prop-notification"
                />
              )}
              {updateError && (
                <InlineNotification
                  lowContrast
                  kind="error"
                  subtitle={"Request to update parameter failed"}
                  title={"Something's Wrong"}
                  data-testid="create-update-team-prop-notification"
                />
              )}
            </ModalBody>
            <ModalFooter>
              <Button kind="secondary" type="button" onClick={closeModal}>
                Cancel
              </Button>
              <Button
                data-testid="team-parameter-create-edit-submission-button"
                type="submit"
                disabled={!isValid || loading}
              >
                {isEdit ? (loading ? "Saving..." : "Save") : loading ? "Creating..." : "Create"}
              </Button>
            </ModalFooter>
          </ModalFlowForm>
        );
      }}
    </Formik>
  );
}

export default CreateEditTeamPropertiesModalContent;
