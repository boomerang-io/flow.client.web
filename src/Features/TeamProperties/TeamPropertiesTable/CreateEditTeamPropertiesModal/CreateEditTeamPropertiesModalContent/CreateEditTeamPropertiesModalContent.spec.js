import React from "react";
import CreateEditTeamPropertiesModalContent from "../CreateEditTeamPropertiesModalContent";
import { fireEvent, waitFor } from "@testing-library/react";
import { act } from "react-dom/test-utils";

const mockfn = jest.fn();
const props = {
  closeModal: mockfn,
  cancelRequestRef: { current: mockfn },
  isEdit: true,
  team: { id: "Test-Id" },
  propertyKeys: ["key"],
  property: {
    id: 1,
  },
};

describe("CreateEditTeamPropertiesModalContent --- Snapshot Test", () => {
  test("Capturing Snapshot of CreateEditTeamPropertiesModalContent", () => {
    const { baseElement } = rtlContextRouterRender(<CreateEditTeamPropertiesModalContent {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});

describe("CreateEditTeamPropertiesModalContent --- RTL Tests", () => {
  test("CreateEditTeamPropertiesModalContent - test if the isActive Toggle appears", () => {
    const newProps = { ...props, isEdit: false };
    const { queryByText } = rtlContextRouterRender(<CreateEditTeamPropertiesModalContent {...newProps} />);

    expect(queryByText(/active/i)).not.toBeInTheDocument();
  });

  test("CreateEditTeamPropertiesModalContent - test the Submit Button state", async () => {
    const newProps = { ...props, isEdit: false };

    const { getByLabelText, getByTestId } = rtlContextRouterRender(
      <CreateEditTeamPropertiesModalContent {...newProps} />
    );
    const valueInputText = getByLabelText(/value/i);
    const labelInputText = getByLabelText(/label/i);
    const keyInputText = getByLabelText(/key/i);

    act(() => {
      fireEvent.change(valueInputText, { target: { value: "testing for value" } });
      fireEvent.change(labelInputText, { target: { value: "testing for label" } });
      fireEvent.change(keyInputText, { target: { value: "rtlTestingKeyTest" } });
    });

    expect(getByTestId("team-parameter-create-edit-submission-button")).toBeEnabled();
    await waitFor(() => {});
  });

  test("CreateEditTeamPropertiesModalContent - test if the form submits", async () => {
    const { getByLabelText, getByTestId } = rtlContextRouterRender(<CreateEditTeamPropertiesModalContent {...props} />);
    const valueInputText = getByLabelText(/value/i);
    const labelInputText = getByLabelText(/label/i);
    const keyInputText = getByLabelText(/key/i);

    expect(valueInputText).toBeInTheDocument();
    expect(labelInputText).toBeInTheDocument();
    expect(keyInputText).toBeInTheDocument();

    act(() => {
      fireEvent.change(valueInputText, { target: { value: "Value Test" } });
      fireEvent.change(labelInputText, { target: { value: "Label Test" } });
      fireEvent.change(keyInputText, { target: { value: "rtlTestingKey" } });
      fireEvent.click(getByTestId("team-parameter-create-edit-submission-button"));
    });

    await waitFor(() => {});
  });

  test("CreateEditTeamPropertiesModalContent - test form reqired validations", async () => {
    const { getByLabelText, findByText, queryByText } = rtlContextRouterRender(
      <CreateEditTeamPropertiesModalContent {...props} />
    );
    const valueInputText = getByLabelText(/value/i);
    const labelInputText = getByLabelText(/label/i);
    const keyInputText = getByLabelText(/key/i);

    expect(queryByText(/Enter a value/i)).toBeNull();
    fireEvent.change(valueInputText, { target: { value: "" } });
    fireEvent.blur(valueInputText);
    const mandatoryValueErr = await findByText(/Enter a value/i);
    expect(mandatoryValueErr).toBeInTheDocument();

    expect(queryByText(/Enter a label/i)).toBeNull();
    fireEvent.change(labelInputText, { target: { value: "" } });
    fireEvent.blur(labelInputText);
    const mandatoryLabelErr = await findByText(/Enter a label/i);
    expect(mandatoryLabelErr).toBeInTheDocument();

    expect(queryByText(/Enter a key/i)).toBeNull();
    fireEvent.change(keyInputText, { target: { value: "" } });
    fireEvent.blur(keyInputText);
    const mandatoryKeyErr = await findByText(/Enter a key/i);
    expect(mandatoryKeyErr).toBeInTheDocument();
    await waitFor(() => {});
  });
});
