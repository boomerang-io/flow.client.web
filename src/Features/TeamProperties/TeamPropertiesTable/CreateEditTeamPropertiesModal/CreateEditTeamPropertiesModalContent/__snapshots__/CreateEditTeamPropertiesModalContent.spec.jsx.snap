// Vitest Snapshot v1

exports[`CreateEditTeamPropertiesModalContent --- Snapshot Test > Capturing Snapshot of CreateEditTeamPropertiesModalContent 1`] = `
<body
  id="app"
>
  <div>
    <form
      class="cds--bmrg-modal-flow-form"
    >
      <div
        aria-label="inputs"
        class="cds--modal-content"
      >
        <div
          class="cds--bmrg-text-input"
        >
          <div
            class="cds--form-item cds--text-input-wrapper"
          >
            <div
              class="cds--text-input__label-wrapper"
            >
              <label
                class="cds--label"
                for="key"
              >
                <div
                  class="cds--bmrg-text-input__label"
                >
                  <div>
                    Key
                  </div>
                </div>
              </label>
            </div>
            <div
              class="cds--text-input__field-outer-wrapper"
            >
              <div
                class="cds--text-input__field-wrapper"
              >
                <input
                  class="cds--text-input cds--text-input--md"
                  id="key"
                  name="key"
                  placeholder="Key"
                  title="Key"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="cds--bmrg-text-input"
        >
          <div
            class="cds--form-item cds--text-input-wrapper"
          >
            <div
              class="cds--text-input__label-wrapper"
            >
              <label
                class="cds--label"
                for="label"
              >
                <div
                  class="cds--bmrg-text-input__label"
                >
                  <div>
                    Label
                  </div>
                </div>
              </label>
            </div>
            <div
              class="cds--text-input__field-outer-wrapper"
            >
              <div
                class="cds--text-input__field-wrapper"
              >
                <input
                  class="cds--text-input cds--text-input--md"
                  id="label"
                  name="label"
                  placeholder="Label"
                  title="Label"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="cds--bmrg-text-input"
        >
          <div
            class="cds--form-item cds--text-input-wrapper"
          >
            <div
              class="cds--text-input__label-wrapper"
            >
              <label
                class="cds--label"
                for="description"
              >
                <div
                  class="cds--bmrg-text-input__label"
                >
                  <div>
                    Description (optional)
                  </div>
                </div>
              </label>
            </div>
            <div
              class="cds--text-input__field-outer-wrapper"
            >
              <div
                class="cds--text-input__field-wrapper"
              >
                <input
                  class="cds--text-input cds--text-input--md"
                  id="description"
                  name="description"
                  placeholder="Description"
                  title="Description"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="cds--bmrg-text-input"
        >
          <div
            class="cds--form-item cds--text-input-wrapper"
          >
            <div
              class="cds--text-input__label-wrapper"
            >
              <label
                class="cds--label"
                for="value"
              >
                <div
                  class="cds--bmrg-text-input__label"
                >
                  <div>
                    Value
                  </div>
                </div>
              </label>
            </div>
            <div
              class="cds--text-input__field-outer-wrapper"
            >
              <div
                class="cds--text-input__field-wrapper"
              >
                <input
                  class="cds--text-input cds--text-input--md"
                  id="value"
                  name="value"
                  placeholder="Value"
                  title="Value"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="cds--bmrg-toggle__container"
        >
          <div
            class="cds--bmrg-toggle --vertical"
          >
            <div
              class="cds--bmrg-toggle__title"
            >
              <label
                class="cds--label"
                for="secured-team-parameters-toggle"
                id="secured-team-parameters-toggle-label"
                style="margin-bottom: 0px;"
              >
                Secured
              </label>
            </div>
            <div
              class="cds--toggle cds--toggle--disabled"
            >
              <button
                aria-checked="false"
                aria-labelledby="secured-team-parameters-toggle-label"
                class="cds--toggle__button"
                data-testid="secured-team-parameters-toggle"
                disabled=""
                id="secured-team-parameters-toggle"
                name="secured"
                role="switch"
                type="button"
              />
              <label
                class="cds--toggle__label"
                for="secured-team-parameters-toggle"
              >
                <span
                  class="cds--toggle__label-text"
                />
                <div
                  class="cds--toggle__appearance"
                >
                  <div
                    class="cds--toggle__switch"
                  />
                  <span
                    aria-hidden="true"
                    class="cds--toggle__text"
                  />
                </div>
              </label>
            </div>
          </div>
          <div
            class="cds--form__helper-text"
            style="margin-bottom: 0px;"
          >
            Once a parameter is securely created - you will not be able to make it unsecure
          </div>
        </div>
      </div>
      <div
        class="cds--modal-footer cds--btn-set"
      >
        <button
          class="cds--btn cds--btn--secondary"
          tabindex="0"
          type="button"
        >
          Cancel
        </button>
        <button
          class="cds--btn cds--btn--primary"
          data-testid="team-parameter-create-edit-submission-button"
          tabindex="0"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</body>
`;
